server/src/main/java/org/opensearch/index/translog/transfer/FileSnapshot.java

```diff
@@ -0,0 +1,223 @@
+import org.opensearch.common.io.stream.InputStreamStreamInput;
+import org.opensearch.core.internal.io.IOUtils;
+import org.opensearch.index.translog.BufferedChecksumStreamInput;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Objects;
+
+/**
+ * Snapshot of a single file that gets transferred
+ *
+ * @opensearch.internal
+ */
+public class FileSnapshot implements Closeable {
+
+    private final String name;
+    @Nullable
+    private final FileChannel fileChannel;
+    @Nullable
+    private Path path;
+    @Nullable
+    private byte[] content;
+
+    public FileSnapshot(Path path) throws IOException {
+        Objects.requireNonNull(path);
+        this.name = path.getFileName().toString();
+        this.path = path;
+        this.fileChannel = FileChannel.open(path, StandardOpenOption.READ);
+    }
+
+    public FileSnapshot(String name, byte[] content) throws IOException {
+        Objects.requireNonNull(name);
+        this.name = name;
+        this.content = content;
+        this.fileChannel = null;
+    }
+
+    public Path getPath() {
+        return path;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public long getContentLength() throws IOException {
+        return fileChannel == null ? content.length : fileChannel.size();
+    }
+
+    public InputStream inputStream() throws IOException {
+        return fileChannel != null
+            ? new BufferedChecksumStreamInput(
+                new InputStreamStreamInput(Channels.newInputStream(fileChannel), fileChannel.size()),
+                path.toString()
+            )
+            : new BufferedChecksumStreamInput(new BytesStreamInput(content), name);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, content, path);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        FileSnapshot other = (FileSnapshot) o;
+        return Objects.equals(this.name, other.name)
+            && Objects.equals(this.content, other.content)
+            && Objects.equals(this.path, other.path);
+    }
+
+    @Override
+    public String toString() {
+        return new StringBuilder("FileInfo [").append(" name = ")
+            .append(name)
+            .append(", path = ")
+            .append(path.toUri())
+            .append("]")
+            .toString();
+    }
+
+    @Override
+    public void close() throws IOException {
+        IOUtils.close(fileChannel);
+    }
+
+    /**
+     * Snapshot of a single file with primary term that gets transferred
+     *
+     * @opensearch.internal
+     */
+    public static class TransferFileSnapshot extends FileSnapshot {
```

Comment:
Can all these nested classes be final?

Reply:
This one is still being extended, have added final to other ones